dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs

```diff
@@ -382,34 +382,28 @@ private Task<string> GetSystemMessage(SKContext context)
 
     private Task<string> GetNextStepCompletion(List<SystemStep> stepsTaken, ChatHistory chatHistory, IAIService aiService, int startingMessageCount, CancellationToken token)
     {
-        var tokenCount = this.GetChatHistoryTokens(chatHistory);
+        var skipStart = startingMessageCount;
+        var skipCount = 0;
 
-        var preserveFirstNSteps = 0;
-        var removalIndex = (startingMessageCount) + preserveFirstNSteps;
-        var messagesRemoved = 0;
         string? originalThought = null;
-        while (tokenCount >= this.Config.MaxTokens && chatHistory.Messages.Count > removalIndex)
-        {
-            // something needs to be removed.
-            if (string.IsNullOrEmpty(originalThought))
-            {
-                originalThought = stepsTaken[0].Thought;
-            }
 
-            // Update message history
-            chatHistory.AddAssistantMessage($"{Thought} {originalThought}");
-            preserveFirstNSteps++;
-            chatHistory.AddAssistantMessage("... I've removed some of my previous work to make room for the new stuff ...");
-            preserveFirstNSteps++;
+        var tokenCount = this.GetChatHistoryTokens(chatHistory);
+        while (tokenCount >= this.Config.MaxTokens && chatHistory.Count > skipStart)
+        {
+            originalThought = $"{Thought} {stepsTaken.FirstOrDefault()?.Thought}";
+            tokenCount = this.GetChatHistoryTokens(chatHistory, $"{originalThought}\n{TrimMessage}", skipStart, ++skipCount);
+        }
 
-            removalIndex = (startingMessageCount) + preserveFirstNSteps;
+        var reducedChatHistory = new ChatHistory();
+        reducedChatHistory.AddRange(chatHistory.Where((m, i) => i < skipStart || i >= skipStart + skipCount));
 
-            chatHistory.Messages.RemoveAt(removalIndex);
-            tokenCount = this.GetChatHistoryTokens(chatHistory);
-            messagesRemoved++;
+        if (skipCount > 0)
+        {
+            reducedChatHistory.InsertMessage(skipStart, AuthorRole.Assistant, TrimMessage);
+            reducedChatHistory.InsertMessage(skipStart, AuthorRole.Assistant, TrimMessage);
```

Comment:
Is it expected to insert the same message two times?

Reply:
Yikes, how in the merge resolution did that happen. Thanks!